pop_tools/budget.py

```diff
@@ -0,0 +1,399 @@
+    mask = grid_f.REGION_MASK.rename('mask')
+    return xr.merge([area, vol, z, mask])
+
+
+def _compute_kmax(budget_depth, z):
+    """Compute the k-index of the maximum budget depth
+
+    Parameters
+    ----------
+    budget_depth : int
+      Maximum depth for control volume in meters
+    z : `xarray.DataArray`
+      Grid cell depths from POP grid
+
+    Returns
+    -------
+    kmax : int
+      k-index for maximum depth of budget calculation
+    """
+    kmax = (z <= budget_depth).argmin() - 1
+    kmax = kmax.values
+    # This only occurs when the full depth is requested (or `budget_depth` is greater than the maximum POP depth`)
+    if kmax == -1:
+        kmax = None
+    return kmax
+
+
+def _convert_to_tendency(ds, grid, sign=1, kmax=None, force_area=False):
+    """Converts from volume- and area-dependent units to tendencies
+
+    Parameters
+    ----------
+    ds : `xarray.Dataset`
+      Main dataset containing all budget-relevant variables
+    grid : `xarray.Dataset`
+      Pre-processed POP grid with area, volume, z, and REGION_MASK
+    sign : int, optional
+      If -1, reverse the sign of the term being computed
+    kmax : int, optional
+      k-index to compute the tendency to
+    force_area : bool, optional
+      If True, normalize by area (used for DIA_IMPVF which typically has a depth component,
+      but requires area-normalization)
+
+    Returns
+    -------
+    ds : `xarray.Dataset`
+      Tendency terms
+    """
+    if force_area:
+        norm = grid.area
+    elif 'z_t' in ds.dims:
+        norm = grid.vol
+        if kmax is not None:
+            norm = norm.isel(z_t=slice(0, kmax + 1))
+    else:
+        norm = grid.area
+    ds = ds * norm * sign
+    ds.attrs['units'] = 'nmol/s'
+    return ds
+
+
+def _convert_units(ds):
+    """Converts from nmol/s to mol/yr and adds attribute label"""
+    ds *= NMOLS_TO_MOLYR
+    ds.attrs['units'] = 'mol/yr'
+    return ds
+
+
+def _subset_slice_mask(ds, var_list=None, mask=None, kmax=None):
+    """Processes the raw dataset to prepare to compute the given budget term.
+
+    Parameters
+    ----------
+    ds : `xarray.Dataset`
+      Main Dataset containing all budget-relevant variables
+    var_list : list
+      Variables to subset from the main dataset
+    mask : `xarray.DataArray`
+      Boolean mask on POP grid
+    kmax : int
+      k-index to slice to
+
+    Returns
+    -------
+    ds : `xarray.Dataset`
+      Dataset subset to relevant variables, sliced to proper depth, and masked
+    """
+    ds = ds[var_list]
+    # Checks that the variable has a depth component before slicing.
+    if (kmax is not None) and ('z_t' in ds.dims):
+        ds = ds.isel(z_t=slice(0, kmax + 1))
+    if mask is not None:
+        ds = ds.where(mask)
+    # If output are already in annual format, this is not an intensive task
+    # and returns the same DataArray.
+    ds = ds.groupby('time.year').mean('time').rename({'year': 'time'})
```

Comment:
This forces all output to annual, which breaks if a user is trying to use daily resolution.